[dtensor][4/N] refactor dispatching logic and add propagator

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92126
* #92931
* #90735
* #90734
* __->__ #90733

This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

Differential Revision: [D42876251](https://our.internmc.facebook.com/intern/diff/D42876251)

wanchaol committed:
[dtensor][3/N] refactor dispatching logic and add propagator

This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][3/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][3/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][3/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][3/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][3/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][3/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][3/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/prop.py:
```diff
@@ -0,0 +1,64 @@
+        if sharding_prop_func is None:
+            # step 1. If there's not even one sharding rule
+            # implemented for the operator, we fall back to
+            # local tensor compute, this is wront currently
```
typo: wront

XilunWu commented on torch/distributed/_tensor/prop.py:
```diff
@@ -0,0 +1,64 @@
+            # implemented for the operator, we fall back to
+            # local tensor compute, this is wront currently
+            # we will change the behavior to reshard to full
+            # replicate and do the computatation
```
typo: computatation

XilunWu reviewed (approved):
Great work! Thx for refactoring op dispatching and sharding propagation facilities. some typos to fix.

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -105,120 +105,71 @@ def _reshape_alias(
+    # implementations. Custom operators take the highest priority
+    if custom_dispatch_ops is not None and str(op_call) in custom_dispatch_ops:
+        # dispatch to user defined custom distributed tensor ops
+        return custom_dispatch_ops[str(op_call)](*args, **kwargs)
```
Yes we should deprecate this once we move all ops to use propagation rules

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876251).

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

Differential Revision: [D42876251](https://our.internmc.facebook.com/intern/diff/D42876251)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

Differential Revision: [D42876251](https://our.internmc.facebook.com/intern/diff/D42876251)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][4/N] refactor dispatching logic and add propagator"


This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

Differential Revision: [D42876251](https://our.internmc.facebook.com/intern/diff/D42876251)

[ghstack-poisoned]

fduwjj reviewed (approved):
LGTM
